### PR TITLE
assign "value" for IntegerBinOp node in StringSection when adjusting for 0-based indexing

### LINKED ISSUE #3273: Compile time value for IntegerBinOp when adjusting for 0-based indexing
For the below Fortran code:
```
program string_slice
    implicit none
    character(len=8) :: s = "lfortran"

    print*, "Length of s:", len(s)
    print*, s(2:3)

end program
```

--show-asr:

```bash
(TranslationUnit
    (SymbolTable
        1
        {
            string_slice:
                (Program
                    (SymbolTable
                        2
                        {
                            s:
                                (Variable
                                    2
                                    s
                                    []
                                    Local
                                    (StringConstant
                                        "lfortran"
                                        (Character 1 8 ())
                                    )
                                    (StringConstant
                                        "lfortran"
                                        (Character 1 8 ())
                                    )
                                    Save
                                    (Character 1 8 ())
                                    ()
                                    Source
                                    Public
                                    Required
                                    .false.
                                )
                        })
                    string_slice
                    []
                    [(Print
                        [(StringConstant
                            "Length of s:"
                            (Character 1 12 ())
                        )
                        (StringLen
                            (Var 2 s)
                            (Integer 4)
                            ()
                        )]
                        ()
                        ()
                    )
                    (Print
                        [(StringSection
                            (Var 2 s)
                            (IntegerBinOp            <------ focus on this IntegerBinOp
                                (IntegerConstant 2 (Integer 4))
                                Sub
                                (IntegerConstant 1 (Integer 4))
                                (Integer 4)
                                ()
                            )
                            (IntegerConstant 3 (Integer 4))
                            (IntegerConstant 1 (Integer 4))
                            (Character 1 8 ())
                            (StringConstant
                                "fo"
                                (Character 1 8 ())
                            )
                        )]
                        ()
                        ()
                    )]
                )
        })
    []
)
```

here the adjustment for 0-based indexing happens, we probably might want to evaluate the value as well, instead of empty `()`.

### PR DESCRIPTION
**Pull request by gxyd**
Fixes #3273

### CODE REVIEW

**Comment by gxyd** [comment]
I wonder for the `--show-c` for the program mentioned in the linked issue (i.e.):
```fortran
program string_slice
    implicit none
    character(len=8) :: s = "lfortran"

    print*, "Length of s:", len(s)
    print*, s(2:3)

end program
```

do we output as:
```bash
(lf1)  gxyd@Gauravs-MacBook-Air  ~/OpenSource/lfortran   main  ./src/bin/lfortran ff.f90 --show-c

#include <stdlib.h>
#include <stdbool.h>
#include <stdio.h>
#include <string.h>
#include <lfortran_intrinsics.h>




// Implementations
int main(int argc, char* argv[])
{
    _lpython_set_argv(argc, argv);
    static char * s = "lfortran";
    printf("%s%s%d\n", "Length of s:", " ", strlen(s));
    printf("%s\n", _lfortran_str_slice(s, 2 - 1, 3, 1, true, true));  <---- this line focus
    return 0;
}
```

i.e. should the index be `2 - 1` or evaluated value (i.e. `1`)?

File: src/lfortran/semantics/ast_common_visitor.h [hunk 1]
```diff
@@ -3203,12 +3203,22 @@ class CommonVisitor : public AST::BaseVisitor<Derived> {
...
+                        ASR::expr_t* a_value = nullptr;
+                        if (args[0].m_left -> type == ASR::exprType::IntegerConstant) {
+                            int64_t a = ASR::down_cast<ASR::IntegerConstant_t>(
+                                            ASRUtils::expr_value(args[0].m_left))->m_n - offset;
+                            a_value = ASRUtils::EXPR((ASR::make_IntegerConstant_t(al, loc,
+                                                    a, int_type)));
+                        }
                         l = ASRUtils::EXPR(ASR::make_IntegerBinOp_t(al, loc,
-                            args[0].m_left, ASR::binopType::Sub, const_1, int_type, nullptr));
+                            args[0].m_left, ASR::binopType::Sub, const_1, int_type, a_value));
```
Comment (certik) [inline src/lfortran/semantics/ast_common_visitor.h:3221]: That's one way to do it. However for such a common operation I think we should use the "builder" interface, which does this automatically. Currently the interface is at https://github.com/lfortran/lfortran/blob/dca110d46606d068dbca0fb334b4bf612ab4011d/src/libasr/pass/intrinsic_function_registry.h#L220. We need to refactor it.

For now, if we don't already have it somewhere, can you create a function that creates `make_IntegerBinOp_t` but takes care of the compile time value automatically?

And put it into `src/libasr/asr_utils.h`.

**Comment by certik** [comment]
I think that this is fine to merge. We can create issues for the other refactoring and fix that later.

**Comment by certik** [comment]
@gxyd let me know how you want to proceed.

**Comment by gxyd** [comment]
Ok @certik , let's go ahead with the merge. I've squashed the commits into one, and kept the commit message intact for them.

**Review by certik — APPROVED** [review]

**Comment by certik** [comment]
Thanks!

**Comment by certik** [comment]
@gxyd this PR committed a 78MB binary into git. I didn't notice it before merging. I just realized it... Trying to figure out a fix in https://github.com/lfortran/lfortran/pull/3577.

**Pull request merged**